File: sdks/go/client_test.go
package reconpro

import (
        "context"
        "errors"
        "os"
        "path/filepath"
        "strings"
        "testing"
        "time"
)

// Hermetic tests: every CLI-backed case runs against a tiny shell-script
// stub binary (the same pattern the SDKs CI workflow uses for the real CLI).
// No reconpro installation is required for `go test ./...`.

// cannedUnreachable is byte-faithful to `reconpro scan <x>.invalid --json`
// on CLI 11.2.0 (plus an unknown "future_field" to pin forward-compat).
const cannedUnreachable = `{
  "target": "nonexistent-target-blah.invalid",
  "modules_run": [],
  "total_findings": 1,
  "severity_counts": {"info": 1},
  "total_score": 0,
  "grade": "U",
  "findings": [
    {
      "title": "Target unreachable — scan not performed",
      "severity": "info",
      "category": "target_validation",
      "module": "engine",
      "description": "honest no-result: no security claims are made",
      "evidence": "state=UNREACHABLE_TARGET; DNS resolution failed",
      "asset": "nonexistent-target-blah.invalid",
      "points_deducted": 0,
      "remediation": "Verify the target hostname is correct and authorized",
      "dread_score": 0.0,
      "confidence": 0.95,
      "verification_state": "UNREACHABLE_TARGET"
    }
  ],
  "target_validation": {
    "state": "UNREACHABLE_TARGET",
    "dns_resolved": false,
    "reachable": false,
    "http_ok": false,
    "tls_valid": null,
    "details": {"host": "nonexistent-target-blah.invalid"},
    "checked_at": "2026-09-18T00:35:16+00:00"
  },
  "scan_metadata": {
    "scanner": "reconpro",
    "started_at": "2026-09-18T00:35:16+00:00",
    "duration_s": 0.01,
    "result": "unreachable_target_short_circuit"
  },
  "future_field": {"anything": ["the", "CLI", "may", "add"]}
}`

// cannedLegacy is an old-CLI result WITHOUT the truth-layer fields — the
// SDK must still parse it (all new fields optional/nil).
const cannedLegacy = `{"target":"example.com","grade":"B","total_findings":1,
 "findings":[{"title":"Old finding","severity":"low"}]}`

// cannedLocal is a local scan (audit/dev): target_validation is null.
const cannedLocal = `{"target":"","modules_run":["dev"],"total_findings":0,
 "severity_counts":{},"total_score":100,"grade":"A+","findings":[],
 "target_validation":null,
 "scan_metadata":{"scanner":"reconpro","started_at":"t","duration_s":0.1,"result":"ok"}}`

func writeStub(t *testing.T, body string) string {
        t.Helper()
        path := filepath.Join(t.TempDir(), "reconpro")
        if err := os.WriteFile(path, []byte("#!/bin/sh\n"+body), 0o755); err != nil {
                t.Fatalf("write stub: %v", err)
        }
        return path
}

func stubClient(t *testing.T, body string) *Client {
        t.Helper()
        return &Client{BinaryPath: writeStub(t, body), Timeout: 10 * time.Second}
}

func TestVersionWithStub(t *testing.T) {
        client := stubClient(t, `echo "ReconPro 11.2.0"`)
        version, err := client.Version(context.Background())
        if err != nil {
                t.Fatalf("Version() error: %v", err)
        }
        if version != "ReconPro 11.2.0" {
                t.Fatalf("unexpected version %q", version)
        }
}

func TestScanTruthLayerContract(t *testing.T) {
        client := stubClient(t, "cat <<'EOF'\n"+cannedUnreachable+"\nEOF\n")
        result, err := client.Scan(context.Background(), "nonexistent-target-blah.invalid")
        if err != nil {
                t.Fatalf("Scan() error: %v", err)
        }
        if result.Grade != "U" {
                t.Fatalf("grade = %q, want U", result.Grade)
        }
        if result.TotalFindings != 1 || result.TotalScore != 0 {
                t.Fatalf("totals = %d/%v, want 1/0", result.TotalFindings, result.TotalScore)
        }
        if len(result.ModulesRun) != 0 {
                t.Fatalf("modules_run = %v, want empty", result.ModulesRun)
        }
        if !result.Unreachable() {
                t.Fatalf("Unreachable() = false, want true")
        }
        tv := result.TargetValidation
        if tv == nil {
                t.Fatal("TargetValidation is nil")
        }
        if tv.State != StateUnreachableTarget {
                t.Fatalf("state = %q, want %q", tv.State, StateUnreachableTarget)
        }
        if tv.DnsResolved == nil || *tv.DnsResolved {
                t.Fatalf("dns_resolved = %v, want false", tv.DnsResolved)
        }
        if len(tv.Details) == 0 {
                t.Fatal("details should decode as non-empty raw JSON")
        }
        if len(result.Findings) != 1 {
                t.Fatalf("findings = %d, want 1", len(result.Findings))
        }
        f := result.Findings[0]
        if f.Severity != "info" { // HIGH impossible by construction
                t.Fatalf("severity = %q, want info", f.Severity)
        }
        if f.Confidence == nil || *f.Confidence != 0.95 {
                t.Fatalf("confidence = %v, want 0.95", f.Confidence)
        }
        if f.VerificationState == nil || *f.VerificationState != StateUnreachableTarget {
                t.Fatalf("verification_state = %v, want %q", f.VerificationState, StateUnreachableTarget)
        }
        meta := result.ScanMetadata
        if meta == nil {
                t.Fatal("ScanMetadata is nil")
        }
        if meta.Scanner != "reconpro" || meta.Result != "unreachable_target_short_circuit" {
                t.Fatalf("metadata = %+v", meta)
        }
        if meta.DurationS != 0.01 {
                t.Fatalf("duration_s = %v, want 0.01", meta.DurationS)
        }
        if result.Raw["future_field"] == nil {
                t.Fatal("Raw should keep unknown fields for forward compatibility")
        }
}

func TestScanLegacyResultWithoutTruthLayerFields(t *testing.T) {
        client := stubClient(t, "cat <<'EOF'\n"+cannedLegacy+"\nEOF\n")
        result, err := client.Scan(context.Background(), "example.com")
        if err != nil {
                t.Fatalf("Scan() error: %v", err)
        }
        if result.Grade != "B" {
                t.Fatalf("grade = %q, want B", result.Grade)
        }
        if result.TargetValidation != nil {
                t.Fatalf("TargetValidation = %+v, want nil", result.TargetValidation)
        }
        if result.ScanMetadata != nil {
                t.Fatalf("ScanMetadata = %+v, want nil", result.ScanMetadata)
        }
        if len(result.Findings) != 1 {
                t.Fatalf("findings = %d, want 1", len(result.Findings))
        }
        f := result.Findings[0]
        if f.Confidence != nil || f.VerificationState != nil {
                t.Fatalf("confidence/verification_state = %v/%v, want nil/nil",
                        f.Confidence, f.VerificationState)
        }
}

func TestLocalScanHasNilTargetValidation(t *testing.T) {
        client := stubClient(t, "cat <<'EOF'\n"+cannedLocal+"\nEOF\n")
        result, err := client.Audit(context.Background())
        if err != nil {
                t.Fatalf("Audit() error: %v", err)
        }
        if result.Grade != "A+" {
                t.Fatalf("grade = %q, want A+", result.Grade)
        }
        if result.TargetValidation != nil || result.Unreachable() {
                t.Fatalf("local scan must have nil target_validation, got %+v",
                        result.TargetValidation)
        }
        if len(result.ModulesRun) != 1 || result.ModulesRun[0] != "dev" {
                t.Fatalf("modules_run = %v, want [dev]", result.ModulesRun)
        }
}

func TestNonZeroExitReturnsCommandError(t *testing.T) {
        client := stubClient(t, `echo "usage: boom" >&2
exit 2`)
        _, err := client.Version(context.Background())
        if err == nil {
                t.Fatal("expected error for exit 2")
        }
        var cmdErr *CommandError
        if !errors.As(err, &cmdErr) {
                t.Fatalf("expected *CommandError, got %T: %v", err, err)
        }
        if cmdErr.ExitCode != ExitUsageError {
                t.Fatalf("ExitCode = %d, want %d", cmdErr.ExitCode, ExitUsageError)
        }
        if !strings.Contains(cmdErr.Stderr, "boom") {
                t.Fatalf("stderr = %q, want it to contain captured text", cmdErr.Stderr)
        }
}

func TestInvalidJSONReturnsJSONError(t *testing.T) {
        client := stubClient(t, `echo "this is not json"`)
        _, err := client.Scan(context.Background(), "example.com")
        if err == nil {
                t.Fatal("expected error for invalid JSON")
        }
        var jsonErr *JSONError
        if !errors.As(err, &jsonErr) {
                t.Fatalf("expected *JSONError, got %T: %v", err, err)
        }
        if !strings.Contains(jsonErr.Details, "this is not json") {
                t.Fatalf("details = %q, want stdout head", jsonErr.Details)
        }
}

func TestTimeoutReturnsTimeoutError(t *testing.T) {
        client := stubClient(t, `sleep 5`)
        client.Timeout = 150 * time.Millisecond
        _, err := client.Version(context.Background())
        if err == nil {
                t.Fatal("expected timeout error")
        }
        var timeoutErr *TimeoutError
        if !errors.As(err, &timeoutErr) {
                t.Fatalf("expected *TimeoutError, got %T: %v", err, err)
        }
        if timeoutErr.Timeout != 150*time.Millisecond {
                t.Fatalf("timeout = %s, want 150ms", timeoutErr.Timeout)
        }
}

func TestBinaryNotFound(t *testing.T) {
        client := &Client{BinaryPath: "/nonexistent/reconpro-xyz", Timeout: time.Second}
        _, err := client.Version(context.Background())
        if err == nil {
                t.Fatal("expected error for nonexistent binary")
        }
        if !strings.Contains(err.Error(), "failed to start") {
                t.Fatalf("expected spawn error mentioning the binary, got %v", err)
        }
}

func TestScanRejectsShellMetacharacters(t *testing.T) {
        client := &Client{BinaryPath: "/x/reconpro"} // never spawned
        if _, err := client.Scan(context.Background(), "example.com; rm -rf /"); err == nil {
                t.Fatal("expected client-side rejection of shell metacharacters")
        } else if !strings.Contains(err.Error(), "metacharacters") {
                t.Fatalf("unexpected error: %v", err)
        }
}

func TestBuildArgsUsesArgumentList(t *testing.T) {
        client := &Client{BinaryPath: "/x/reconpro"}
        args := client.buildArgs("scan", "example.com", "--json")
        want := []string{"/x/reconpro", "scan", "example.com", "--json"}
        if len(args) != len(want) {
                t.Fatalf("buildArgs = %v, want %v", args, want)
        }
        for i := range want {
                if args[i] != want[i] {
                        t.Fatalf("buildArgs = %v, want %v", args, want)
                }
        }
}

func TestExportValidatesFormat(t *testing.T) {
        client := &Client{BinaryPath: "/x/reconpro"} // never spawned
        if _, err := client.Export(context.Background(), "/tmp/report", "pdf"); err == nil {
                t.Fatal("expected error for unsupported export format")
        }
}

func TestNewClientReadsEnv(t *testing.T) {
        t.Setenv("RECONPRO_BINARY", "/from/env/reconpro")
        client := NewClient()
        if client.BinaryPath != "/from/env/reconpro" {
                t.Fatalf("BinaryPath = %q, want /from/env/reconpro", client.BinaryPath)
        }
        if client.Timeout != DefaultTimeout {
                t.Fatalf("Timeout = %s, want %s", client.Timeout, DefaultTimeout)
        }
}

func TestDevValidatesPath(t *testing.T) {
        client := &Client{BinaryPath: "/x/reconpro"} // never spawned
        if _, err := client.Dev(context.Background(), "   "); err == nil {
                t.Fatal("expected error for empty path")
        }
}

File: sdks/typescript/tests/client.test.ts
/**
 * Test suite for reconpro-sdk (TypeScript, Stable tier — see sdks/POLICY.md).
 *
 * Hermetic: every CLI-backed case runs against the stub binary
 * `tests/fixtures/reconpro-stub.mjs` (canned JSON byte-faithful to
 * `reconpro scan <x>.invalid --json` on CLI 11.2.0, plus an unknown
 * "future_field" to pin forward-compat). Error paths are exercised through
 * the real public SDK API via the stub's documented magic targets.
 *
 * Run:  bun test          (or: bun test tests/client.test.ts)
 */

import { describe, expect, test } from "bun:test";
import { chmod, mkdtemp, readFile, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import {
  ReconProClient,
  ReconProError,
  UNREACHABLE_TARGET,
  TARGET_STATES,
  validateTarget,
  normalizeScanResult,
  EXPORT_FORMATS,
  DEFAULT_BINARY,
  DEFAULT_TIMEOUT_MS,
} from "../src/index.js";

const TEST_DIR = dirname(fileURLToPath(import.meta.url));
export const STUB_PATH = join(TEST_DIR, "fixtures", "reconpro-stub.mjs");

const stubClient = (options = {}) =>
  new ReconProClient({ binaryPath: STUB_PATH, timeoutMs: 15_000, ...options });

/** Minimal sh-script stub (for quirks the fixture does not model). */
async function shStub(body: string): Promise<string> {
  const dir = await mkdtemp(join(tmpdir(), "reconpro-sdk-ts-"));
  const path = join(dir, "reconpro-stub");
  await writeFile(path, `#!/bin/sh\n${body}`, "utf8");
  await chmod(path, 0o755);
  return path;
}

/** Await a promise that must reject with a ReconProError; return it typed. */
async function expectSdkError(
  fn: () => Promise<unknown>,
): Promise<ReconProError> {
  try {
    await fn();
  } catch (e) {
    expect(e).toBeInstanceOf(ReconProError);
    return e as ReconProError;
  }
  throw new Error("expected the promise to reject with a ReconProError");
}

// ─────────────────────────── construction & config ─────────────────────────

describe("client construction", () => {
  test("constructor options take precedence; env vars are honoured", () => {
    const before = process.env.RECONPRO_BINARY;
    process.env.RECONPRO_BINARY = "/from/env/reconpro";
    try {
      expect(new ReconProClient().binaryPath).toBe("/from/env/reconpro");
      expect(
        new ReconProClient({ binaryPath: "/explicit/reconpro" }).binaryPath,
      ).toBe("/explicit/reconpro");
    } finally {
      if (before === undefined) delete process.env.RECONPRO_BINARY;
      else process.env.RECONPRO_BINARY = before;
    }
  });

  test("defaults: binary name + 300 s timeout", () => {
    const client = new ReconProClient({ binaryPath: "/x/reconpro" });
    expect(client.binaryPath).toBe("/x/reconpro");
    expect(DEFAULT_BINARY).toBe("reconpro");
    expect(DEFAULT_TIMEOUT_MS).toBe(300_000);
  });

  test("buildCommand is an argument list, never a shell string", () => {
    const client = new ReconProClient({ binaryPath: "/x/reconpro" });
    expect(client.buildCommand(["scan", "t", "--json"])).toEqual([
      "/x/reconpro",
      "scan",
      "t",
      "--json",
    ]);
  });
});

// ─────────────────────────── truth-layer contract ─────────────────────────

describe("scan — truth-layer contract (hermetic, stub-backed)", () => {
  test("unreachable *.invalid target returns the honest no-result", async () => {
    const result = await stubClient().scan("nonexistent-target-blah.invalid");

    expect(result.grade).toBe("U");
    expect(result.totalFindings).toBe(1);
    expect(result.totalScore).toBe(0);
    expect(result.modulesRun).toEqual([]);
    expect(result.severityCounts).toEqual({ info: 1 });
    expect(result.unreachable).toBe(true);

    const tv = result.targetValidation;
    expect(tv).not.toBeNull();
    expect(tv!.state).toBe(UNREACHABLE_TARGET);
    expect(tv!.dnsResolved).toBe(false);
    expect(tv!.reachable).toBe(false);
    expect(tv!.httpOk).toBe(false);
    expect(tv!.tlsValid).toBeNull();
    expect(tv!.checkedAt).toBeTypeOf("string");

    expect(result.findings).toHaveLength(1);
    const f = result.findings[0];
    expect(f.severity).toBe("info"); // HIGH impossible by construction
    expect(f.category).toBe("target_validation");
    expect(f.pointsDeducted).toBe(0);
    expect(f.verificationState).toBe(UNREACHABLE_TARGET);
    expect(f.confidence).toBeCloseTo(0.95);
    expect(f.dreadScore).toBe(0);

    const meta = result.scanMetadata;
    expect(meta).not.toBeNull();
    expect(meta!.scanner).toBe("reconpro");
    expect(meta!.result).toBe("unreachable_target_short_circuit");
    expect(meta!.durationS).toBeCloseTo(0.01);
    expect(meta!.startedAt).toBeTypeOf("string");

    // Unknown keys stay reachable via raw (forward compatibility).
    expect(
      (result.raw.future_field as { anything: string[] }).anything[0],
    ).toBe("the");
  });

  test("legacy result without truth-layer fields still parses", async () => {
    const legacy = {
      target: "example.com",
      grade: "B",
      total_findings: 1,
      findings: [{ title: "Old finding", severity: "low" }],
    };
    const result = normalizeScanResult(legacy);
    expect(result.grade).toBe("B");
    expect(result.targetValidation).toBeNull();
    expect(result.scanMetadata).toBeNull();
    expect(result.unreachable).toBe(false);
    expect(result.findings[0].confidence).toBeNull();
    expect(result.findings[0].verificationState).toBeNull();
  });

  test("null target_validation (local scans) is tolerated", () => {
    const result = normalizeScanResult({ target: "x", target_validation: null });
    expect(result.targetValidation).toBeNull();
    expect(result.unreachable).toBe(false);
  });

  test("vibesec() returns the same typed contract", async () => {
    const result = await stubClient().vibesec("nonexistent-target-blah.invalid");
    expect(result.grade).toBe("U");
    expect(result.targetValidation!.state).toBe(UNREACHABLE_TARGET);
  });

  test("audit()/dev()/doctor() return local typed results", async () => {
    const client = stubClient();
    for (const result of [await client.audit(), await client.dev(), await client.doctor()]) {
      expect(result.modulesRun).toEqual(["dev"]);
      expect(result.targetValidation).toBeNull();
      expect(result.grade).toBe("A+");
    }
  });
});

// ─────────────────────────────── error contract ───────────────────────────

describe("structured errors", () => {
  test("BIN_NOT_FOUND when the binary cannot be spawned", async () => {
    const client = new ReconProClient({
      binaryPath: "/nonexistent/reconpro-xyz",
      timeoutMs: 5_000,
    });
    const err = await expectSdkError(() => client.version());
    expect(err.code).toBe("BIN_NOT_FOUND");
  });

  test("TIMEOUT kills the process and rejects", async () => {
    const client = stubClient({ timeoutMs: 300 });
    const err = await expectSdkError(() => client.scan("sleepy-target.invalid"));
    expect(err.code).toBe("TIMEOUT");
  });

  test("NON_ZERO_EXIT carries exit code + captured stderr", async () => {
    const err = await expectSdkError(() =>
      stubClient().scan("make-it-fail.invalid"),
    );
    expect(err.code).toBe("NON_ZERO_EXIT");
    expect(err.exitCode).toBe(2);
    expect(err.stderr).toContain("boom");
  });

  test("INVALID_JSON when stdout is not JSON (exit 0)", async () => {
    const err = await expectSdkError(() =>
      stubClient().scan("garbage.invalid"),
    );
    expect(err.code).toBe("INVALID_JSON");
    expect(err.stderr).toContain("this is not json"); // stdout head for diagnostics
  });

  test("INVALID_JSON when stdout is a JSON array, not an object", async () => {
    const stub = await shStub("echo '[1, 2, 3]'");
    const err = await expectSdkError(() =>
      new ReconProClient({ binaryPath: stub, timeoutMs: 5_000 }).audit(),
    );
    expect(err.code).toBe("INVALID_JSON");
  });

  test("INVALID_TARGET for shell metacharacters (client-side)", () => {
    expect(() => validateTarget("example.com; rm -rf /")).toThrow(ReconProError);
    expect(() => validateTarget("   ")).toThrow(ReconProError);
    const err = (() => {
      try {
        return validateTarget("$(whoami)");
      } catch (e) {
        expect(e).toBeInstanceOf(ReconProError);
        return e as ReconProError;
      }
    })();
    expect(err.code).toBe("INVALID_TARGET");
  });

  test("UNSUPPORTED_FORMAT for bad export formats", async () => {
    await expect(stubClient().exportReport("pdf" as never)).rejects.toMatchObject({
      code: "UNSUPPORTED_FORMAT",
    });
  });
});

// ────────────────────────────── output files ──────────────────────────────

describe("-o output file round-trip", () => {
  test("JSON is read back from the file (CLI writes only there)", async () => {
    const dir = await mkdtemp(join(tmpdir(), "reconpro-sdk-ts-out-"));
    const out = join(dir, "scan.json");
    const result = await stubClient().scan("nonexistent-target-blah.invalid", {
      outputFile: out,
    });
    expect(result.grade).toBe("U");
    expect(result.targetValidation!.state).toBe(UNREACHABLE_TARGET);
    const onDisk = JSON.parse(await readFile(out, "utf8"));
    expect(onDisk.target).toBe("nonexistent-target-blah.invalid");
    expect(onDisk.grade).toBe("U");
  });

  test("missing output file → INVALID_JSON", async () => {
    const stub = await shStub("exit 0"); // claims -o but writes nothing
    const err = await expectSdkError(() =>
      new ReconProClient({ binaryPath: stub, timeoutMs: 5_000 }).scan(
        "example.com",
        { outputFile: "/never-written-by-stub.json" },
      ),
    );
    expect(err.code).toBe("INVALID_JSON");
  });
});

// ──────────────────────────────── export ──────────────────────────────────

describe("exportReport", () => {
  test("writes via the CLI export subcommand and returns the path", async () => {
    const dir = await mkdtemp(join(tmpdir(), "reconpro-sdk-ts-exp-"));
    const out = join(dir, "report.sarif");
    const written = await stubClient().exportReport("sarif", out);
    expect(written).toBe(out);
    const content = await readFile(out, "utf8");
    expect(content).toContain("stub");
  });

  test("appends the format extension when missing", async () => {
    const dir = await mkdtemp(join(tmpdir(), "reconpro-sdk-ts-exp2-"));
    const out = join(dir, "report");
    const written = await stubClient().exportReport("md", out);
    expect(written).toBe(`${out}.md`);
  });

  test("EXPORT_FORMATS covers sarif/md/json/html", () => {
    expect([...EXPORT_FORMATS]).toEqual(["sarif", "md", "json", "html"]);
  });
});

// ───────────────────────────── version ping ───────────────────────────────

describe("version", () => {
  test("returns the CLI version string from stdout", async () => {
    expect(await stubClient().version()).toBe("ReconPro 11.2.0");
  });

  test("usage-error exit 2 surfaces as NON_ZERO_EXIT (sh stub)", async () => {
    const stub = await shStub('echo "usage: boom" >&2\nexit 2\n');
    const err = await expectSdkError(() =>
      new ReconProClient({ binaryPath: stub, timeoutMs: 5_000 }).version(),
    );
    expect(err.code).toBe("NON_ZERO_EXIT");
    expect(err.exitCode).toBe(2);
    expect(err.stderr).toContain("boom");
  });
});

// Re-exported helper contract (kept in sync with the other SDKs' constants).
expect([...TARGET_STATES]).toEqual([
  "VERIFIED_TARGET",
  "PARTIAL_TARGET",
  "UNREACHABLE_TARGET",
]);
